tracer/src/Datadog.Trace/Agent/MessagePack/CachedStringBytes.cs

```diff
@@ -0,0 +1,31 @@
+﻿// <copyright file="CachedStringBytes.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System.Text;
+
+namespace Datadog.Trace.Agent.MessagePack;
+
+internal readonly struct CachedStringBytes
+{
+    public readonly byte[]? Environment;
+
+    public readonly byte[]? ServiceVersion;
+
+    public readonly byte[]? Origin;
+
+    public CachedStringBytes(Encoding encoding, string? environment, string? serviceVersion, string? origin)
+    {
+        Environment = GetBytes(encoding, environment);
+        ServiceVersion = GetBytes(encoding, serviceVersion);
+        Origin = GetBytes(encoding, origin);
+    }
```

Comment:
There is:
```csharp
var env = span.Context.TraceContext?.Environment;

if (!string.IsNullOrWhiteSpace(env))
{
    count++;
    offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _environmentNameBytes);
    offset += MessagePackBinary.WriteString(ref bytes, offset, env);
}
```

For a small trace chunk this could be faster by avoiding the extra `byte[]` allocations, but for larger chunks this code encodes `env` into UTF-8 for every span. I _think_ there's a point where caching the bytes if better than not. 10 spans? 50? I should probably try to figure this out.

Reply:
If we start caching these bytes across traces (as long as they don't change, and they should almost never change), then this is less of an issue.

Reply:
I did a thing with `[ThreadStatic]` like you suggested, so this should be less of an issue now.

Reply:
> There is. In the MsgPack formatter

Oh yeah, I knew that really 😅 🤦‍♂️